core/src/main/scala/chisel3/RelativeOrientation.scala

```diff
@@ -0,0 +1,98 @@
+package chisel3
+
+import chisel3.experimental.DataMirror
+
+
+// Indicates whether the active side is aligned or flipped relative to the active side's root
+sealed trait RelativeOrientation { 
+  def data: Data
+  def waivers: Set[Data]
+  def invert: RelativeOrientation
+  def coerced: Boolean
+  def coerce: RelativeOrientation
+  def alignsWith(o: RelativeOrientation): Boolean = o.alignment == this.alignment
+  def childOrientations: Seq[RelativeOrientation] = data match {
+    case a: Aggregate => a.getElements.map(e => RelativeOrientation.deriveOrientation(e, this))
+    case o => Nil
+  }
+  def alignment: String
+  def swap(d: Data): RelativeOrientation
+  def isWaived: Boolean = waivers.contains(data)
+  def isAgg: Boolean = data.isInstanceOf[Aggregate]
+  def isConsumer: Boolean
+  def errorWord(op: DirectionalConnectionFunctions.ConnectionOperator): String = (isConsumer, op.assignToConsumer, op.assignToProducer, alignment) match {
+    case (true,  true,   _,    "aligned") => "unassigned"
+    case (false, _,      true, "flipped") => "unassigned"
+    case (true,  _,      true, "flipped") => "dangling"
+    case (false, true,   _,    "aligned") => "dangling"
+    case other => "unmatched"
+  }
+}
+object RelativeOrientation {
+  def apply(d: Data, waivers: Set[Data], isConsumer: Boolean): RelativeOrientation = AlignedWithRoot(d, isCoercing(d), waivers, isConsumer)
+  def isCoercing(d: Data): Boolean = {
+    def recUp(x: Data): Boolean = x.binding match {
+      case _ if isCoercing(x)           => true
+      case None                         => false
+      case Some(t: internal.TopBinding) => false
+      case Some(internal.ChildBinding(p))        => recUp(p)
+      case other                        => throw new Exception(s"Unexpected $other! $x, $d")
+    }
+    def isCoercing(d: Data): Boolean = {
```

Comment:
Can you call this something different than the outer function? Personally this is confusing to have the same name